react-web/src/components/ConnectWallet/ConnectWallet.tsx

```diff
@@ -42,21 +47,64 @@ const ConnectWallet = () => {
                 setWalletName(walletName)
                 if (enabledWallet) {
                     const response = await enabledWallet.getChangeAddress()
-                    setAddress(Address.from_bytes(Buffer.from(response, "hex")).to_bech32())
+                    const walletAddr = Address.from_bytes(Buffer.from(response, "hex")).to_bech32()
+                    initiatePrivateWalletSignature(enabledWallet, walletAddr, response)
                 }
             })
         } catch (e) { handleError(e); }
     }
 
     const handleError = (error: any) => {
         if (error.info) {
-            setErrorToast({display: true, message: error.info})
+            setErrorToast({
+                display: true, 
+                message: error.info, 
+                showRetry: error.code === 3})
         } else if (error.response) {
-          setErrorToast({display: true, statusText: error.response.statusText, message: error.response.data || undefined})
+            setErrorToast({
+                display: true, 
+                statusText: error.response.statusText, 
+                message: error.response.data || undefined,
+                showRetry: error.status === 403
+            })
         } else {
           setErrorToast({display: true})
         }
-        setTimeout(() => { setErrorToast({display: false}) }, 3000)
+    }
+
+    const catchError = (err: any) => {
+        handleError(err)
+        setWalletLoading(false)
+        dispatch(logout())
+    }
+
+    const initiatePrivateWalletSignature = (currentWallet: any, walletAddr_bech32: any, walletAddr: string) => {
+        fetchData.get('/server-timestamp').then(async (res) => {
+            const timestamp = res.data;
+            const msgToBeSigned = `Sign this message if you are the owner of the ${walletAddr_bech32} address. \n Timestamp: <<${timestamp}>> \n Expiry: 60 seconds`;
+            try {
+                const {key, signature} = await currentWallet.signData(walletAddr, Buffer.from(msgToBeSigned, 'utf8').toString('hex'))
+                if (key && signature) {
+                    fetchData.post('/login', {
+                        address: walletAddr_bech32,
+                        key: key,
+                        signature: signature
+                    }).then((response: any) => {
+                        if (response.data) {
+                            localStorage.setItem('authToken', response.data)
+                            setAddress(walletAddr_bech32)
+                        }
+                    }).catch(err => {
+                        catchError(err)
+                    })
+                }
+
+            } catch (err) {
+                catchError(err)
+            }
+        }).catch((err) => {
+            catchError(err)
+        })
```

Comment:
- Why not using async/await
- The code is not clean and hard to follow with all these multiple catches. Here is an alternative: 
```suggestion
    const initiatePrivateWalletSignature = async (currentWallet: any, walletAddr_bech32: any, walletAddr: string) => {
      try{
        const timestamp = (await fetchData.get<any,AxiosResponse<number>,any>('/server-timestamp')).data;
        const msgToBeSigned = `Sign this message if you are the owner of the ${walletAddr_bech32} address. \n Timestamp: <<${timestamp}>> \n Expiry: 60 seconds`;
        const {key, signature} = await currentWallet.signData(walletAddr, Buffer.from(msgToBeSigned, 'utf8').toString('hex'))
        if (key && signature) {
          const token = (await fetchData.post<any,AxiosResponse<string>,any>('/login', {
            address: walletAddr_bech32,
            key: key,
            signature: signature
          })).data;
          localStorage.setItem('authToken', token)
          setAddress(walletAddr_bech32)
        }
      } catch (err) {
        catchError(err)
      }
```
- If you agree with the alternative there is still an issue: What do we do when there is no (key/signature), I see you have an if, but there is no else branch, shouldn't the user be notified somehow?

Reply:
Done